File: solutions/javascript/blind-75/array/bestTimeToBuyAndSellStock.js
/*
For each stock price, we compute the max profit using that stock price. To do so, we need to know the minimum stock price that appears BEFORE the current stock price. Thus, we continuously track the minimum stock price so far. Then, at each stock, we compute the max profit using the current stock (which is cur stock price - min price so far) and keep a running variable for our max profit.
*/
const maxProfit = (prices) => {
  let minPriceSoFar = Infinity;
  let maxProfit = 0;

  prices.forEach((price) => {
    const maxProfitUsingCurPrice = price - minPriceSoFar;

    minPriceSoFar = Math.min(minPriceSoFar, price);
    maxProfit = Math.max(maxProfit, maxProfitUsingCurPrice);
  });

  return maxProfit;
};

File: solutions/python/bfs/blind-75/bestTimeToBuyAndSellStock.py
class Solution:
    def maxProfit(self, prices):
        min_price_so_far = float('inf')
        max_profit = 0

        for price in prices:
            cur_max_profit = price - min_price_so_far
            
            if cur_max_profit > max_profit:
                max_profit = cur_max_profit
                
            if price < min_price_so_far:
                min_price_so_far = price
                
        return max_profit